src/Draco.Compiler/Internal/Solver/ConstraintSolver_Utils.cs

```diff
@@ -15,6 +16,22 @@ internal sealed partial class ConstraintSolver
 {
     private readonly record struct OverloadCandidate(FunctionSymbol Symbol, CallScore Score);
 
+    private static bool IsBaseOf(TypeSymbol @base, TypeSymbol derived)
+    {
+        @base = @base.Substitution;
+        @derived = @derived.Substitution;
+
+        if (!@base.IsGroundType || !derived.IsGroundType) throw new InvalidOperationException();
+
+        // NOTE: Duplicate logic from unification
+        // TODO: Can we factor it out?
+        if (@base is NeverTypeSymbol or ErrorTypeSymbol) return true;
+        if (derived is NeverTypeSymbol or ErrorTypeSymbol) return true;
+
+        return SymbolEqualityComparer.Default.Equals(@base, derived)
+            || derived.ImmediateBaseTypes.Any(b => IsBaseOf(@base, b));
```

Comment:
If you included the type itself in the BaseTypes, this would be simpler and non-recursive